recipes/btyacc/all/test_package/conanfile.py

```diff
@@ -1,23 +1,48 @@
-import functools
-import os
-
-from conans import ConanFile, CMake, tools
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import CMake, CMakeDeps, cmake_layout
+from conan.tools.env import VirtualBuildEnv, VirtualRunEnv
 
 
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
-    generators = "cmake", "cmake_find_package_multi"
+    generators = "CMakeToolchain"
+    test_type = "explicit"
+
+    def layout(self):
+        cmake_layout(self)
+
+    # TODO: Only keep logic under if hasattr(self, "settings_build") after conan v2.
+    #       Indeed, v1 pipeline of c3i uses 1 profile, which can't work with
+    #       build_context_activated & build_context_build_modules of CMakeDeps
+
+    def requirements(self):
+        if not hasattr(self, "settings_build"):
+            self.requires(self.tested_reference_str)
+
+    def build_requirements(self):
+        if hasattr(self, "settings_build"):
+            self.tool_requires(self.tested_reference_str)
 
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        cmake = CMake(self)
-        cmake.configure()
-        return cmake
+    def generate(self):
+        if hasattr(self, "settings_build"):
+            VirtualBuildEnv(self).generate()
+            deps = CMakeDeps(self)
+            deps.build_context_activated = ["btyacc"]
+            deps.build_context_build_modules = ["btyacc"]
+            deps.generate()
+        else:
+            if can_run(self):
+                VirtualRunEnv(self).generate(scope="build")
+                deps = CMakeDeps(self)
+                deps.generate()
 
     def build(self):
-        if not tools.cross_building(self):
-            self._configure_cmake().build()
+        if hasattr(self, "settings_build") or can_run(self):
+            cmake = CMake(self)
+            cmake.configure()
+            cmake.build()
 
     def test(self):
-        if not tools.cross_building(self):
-            self._configure_cmake().test()
+        if can_run(self):
+            self.run(f"ctest --output-on-failure -C {self.settings.build_type}", env="conanrun")
```

Comment:
@SpaceIm I don't get anything about this test_package. I know that this code was somehow there, but let's discuss this one.

This recipe is an `application` one, so, IMO, this conanfile should look like:

```python
from conan import ConanFile


class TestPackageConan(ConanFile):
    settings = "os", "arch", "compiler", "build_type"
    generators = "VirtualBuildEnv"
    test_type = "explicit"

    def build_requirements(self):
        self.tool_requires(self.tested_reference_str)

    def test(self):
        self.run("btyacc -b y [THE REST OF THE PARAMETERS HERE]")
```
Asking because I'm sure I'm missing something somewhere...

Reply:
This test package also test that we can consume btyacc with find_package() and that Cmake imported target of executable works as expected, so we can't rely on such naive test.

Reply:
Yeah, but it's only using the `BTYACC_EXECUTABLE` variable from there, so I was wondering if it seemed too much, and could be simpler than that.

Perhaps:

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.1)
project(test_package LANGUAGES C)

find_package(btyacc REQUIRED CONFIG)
if(NOT DEFINED BTYACC_EXECUTABLE)
    message(FATAL_ERROR "BTYACC_EXECUTABLE is not defined")
endif()
```
conanfile
```python
import os

from conan import ConanFile
from conan.tools.cmake import cmake_layout, CMake


class TestPackageConan(ConanFile):
    settings = "os", "arch", "compiler", "build_type"
    generators = "CMakeDeps", "CMakeToolchain", "VirtualBuildEnv"
    test_type = "explicit"

    def layout(self):
        cmake_layout(self)

    def build_requirements(self):
        self.tool_requires(self.tested_reference_str)

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def test(self):
        self.run("btyacc -b y [THE REST OF THE PARAMETERS HERE]")
```
It's only an idea. Testing both CMake and env executable at the same time could be a better fit.

Reply:
It wouldn't be a representative usage. I don't understand what's the problem with current test package. It works and it's a representative test.

Reply:
No no, it's fine for me. I was only thinking of other alternatives to test the same thing, but this one was already there so I have no objections.